### ISSUE TITLE
Trigger Model Events when using @update, @create, @delete

### ISSUE THREAD

**Issue by LRstudentHU** [nuwave/lighthouse#677]
**Is your feature request related to a problem? Please describe.**
Currently, the following mutations will not trigger a model event:
```js
createAccount(
    // Arguments
): Account @create(model: "App\\Models\\Account")

updateAccount(
    id: ID! @rules(apply: ["exists:accounts,id"])
    // Arguments
): Account @update(model: "App\\Models\\Account")

deleteAccount(
    id: ID! @rules(apply: ["exists:accounts,id"])
): Account @delete(model: "App\\Models\\Account")
```
Here, only a specific model is being created, updated or deleted. So it would make sense that using these directives trigger the corresponding model events.

**Describe the solution you'd like**
Using the `@update`, `@delete` or `@create` directives trigger model events.

**Describe alternatives you've considered**
A workaround is using a resolver and manually creating or updating the given model.
This works, but defeats the purpose of having these directives really.

I'm not sure if this is a bug or a feature request.

**Comment by spawnia** [comment]
> I'm not sure if this is a bug or a feature request.

I will put it up for discussion.

My opinion is that the events should definitely trigger, nothing special about mutating the Model through GraphQL that would warrant turning off this useful Laravel feature.

**Comment by LRstudentHU** [comment]
Investigating further, model events actually _do_ fire. Inspecting the `MutationExecutor`, `->save()` is always called on a model instance. I had suspected that a generic `->update` was used, causing the events to not fire.

In my testing, I was under the impression that calling `->save()` on a model _always_ fired the `saving`  and `saved` events. This is actually not the case; Laravel only fires these events when attributes haves actually changed. I haven't tested the relationships extensively, but they should work fine as well.

**Comment by mfn** [comment]
> In my testing, I was under the impression that calling `->save()` on a model _always_ fired the `saving` and `saved` events. This is actually not the case; Laravel only fires these events when attributes haves actually changed. I haven't tested the relationships extensively, but they should work fine as well.

But that's not accurate, either:
- `saving` is **always** fired (it's the event _before_ the save happens), but it is fired for "new" (created) as well as updated models
- `saved` is **only** fired if something was **actually** done (i.e. actual changes, what you mentioned), but again: for both "new" (created) or updated models

And then you've:
- `creating`: only when a new model is going to be inserted
- `created`: when it's inserted
- `updating`: only when an existing model is going to be updated (achtung: only if it "has changed"!)
- `updated`: when it was actually updated (achtung again: only if it "had changed"")

In a nutshell, for
- new model created: `saving`, `creating`, `created`, `saved`
- existing model updated: `saving`, `updating`, `updated`, `saved`
  but `updating`, `updated` and `saved` only if there were actual changes

As long as `->save()` on models is called, these streams of events apply.

(I hope I did get the orders right for all 😄 )

**Comment by spawnia** [comment]
I guess we do call `destroy()` in some parts, which would be the one exception where it would not trigger the lifecycle events.

Open to reworking that for consistency.

**Comment by lorado** [comment]
I am currently developing completely changes history log for my essential models, and in my testing events gets correctly fired. I listen to `created`, `updated` and `deleted` events, as those are actually events that signals changes in DB. There are some problems in relations with additional pivot-data and morph pivot, but they belong to laravel itself. 

I also created issues there, so if you want to follow:

- https://github.com/laravel/framework/issues/29636
- https://github.com/laravel/framework/issues/29631

**Comment by spawnia** [comment]
Right. Actually, `Model::destroy() ` triggers the delete event, so I think there is nothing left to do.
